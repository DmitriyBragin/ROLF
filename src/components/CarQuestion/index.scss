.car-question {
    display: flex;
    flex-direction: column;
    margin-bottom: 40px;
    &__header {
        display: flex;
        flex-direction: row;
        font-weight: 700;
        &_name {
        }
        &_car {
            margin-left: 20px;
            color: #923562;
        }
    }
    &__time {
        color: grey;
        margin-bottom: 10px;
    }
    &__text {
    }
}
