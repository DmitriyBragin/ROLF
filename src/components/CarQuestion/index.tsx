import './index.scss';
import { Question } from '../../utils/types';

export const CarQuestion = ({ question }: { question: Question }) => {
    const transformDate = (): string => {
        if (!question.date) return 'N/A';
        return question.date.split('T').join(' Ð² ');
    };

    return question ? (
        <div className="car-question">
            <div className="car-question__header">
                <div className="car-question__header_name">{question.user}</div>
                <div className="car-question__header_car">
                    {question.carBrand} {question.carModel}
                </div>
            </div>
            <div className="car-question__time">{transformDate()}</div>
            <div className="car-question__text">{question.text}</div>
        </div>
    ) : (
        <></>
    );
};
