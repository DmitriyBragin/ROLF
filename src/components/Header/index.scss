.header {
    color: #3f48cc;
    height: 75px;
    display: flex;
    flex-direction: row;
    flex: 0 0 75px;
    align-items: center;
    font-size: 24px;
    font-weight: 700;
}
