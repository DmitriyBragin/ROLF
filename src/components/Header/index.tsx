import './index.scss';

export const Header = () => {
    return <div className="header">Вопросы по автомобилям</div>
}