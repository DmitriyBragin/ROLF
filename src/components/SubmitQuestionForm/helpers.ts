import { Question } from '../../utils/types';

type FormValidationProps = {
    errors: Question;
    isValid: boolean;
};

type TextLengthRules = {
    minChars: number;
    maxChars: number;
};

type TextRules = {
    lengthRules?: TextLengthRules;
    regexp: RegExp;
};

export const isFilled = (object: Object): boolean => Object.values(object).every(x => x !== '');
export const isNotEmpty = (object: Object): boolean => Object.values(object).some(x => x !== '');

export const getEmptyQuestionObject = (): Question => {
    return {
        user: '',
        email: '',
        carBrand: '',
        carModel: '',
        text: '',
    };
};

const textFollowsRules = (text: string, rules: TextRules): boolean => {
    const invalidChars = text.trim().replaceAll(rules.regexp, '').length;
    if (!rules.lengthRules) return invalidChars > 0;
    return (
        text.trim().length < rules.lengthRules.minChars ||
        text.trim().length > rules.lengthRules.maxChars ||
        invalidChars > 0
    );
};

const validateUserField = (text: string): string => {
    let res: string = '';
    const userRegexp = /[\p{Script=Cyrillic}\p{White_Space}]/gu;
    const userLengthRules: TextLengthRules = {
        minChars: 2,
        maxChars: 30,
    };
    const userTextRules: TextRules = {
        lengthRules: userLengthRules,
        regexp: userRegexp,
    };
    if (!textFollowsRules(text, userTextRules)) {
        res = 'Только русские буквы от 2х до 30 символов';
    }
    return res;
};

const validateCarModelField = (text: string): string => {
    let res: string = '';
    const carModelRegexp = /[\p{Script=Latin}\p{Number}\p{White_Space}]/gu;
    const carModelLengthRules: TextLengthRules = {
        minChars: 4,
        maxChars: 20,
    };
    const carModelTextRules: TextRules = {
        lengthRules: carModelLengthRules,
        regexp: carModelRegexp,
    };
    if (!textFollowsRules(text, carModelTextRules)) {
        res = 'Только латинские буквы и цифры от 4х до 20 символов';
    }
    return res;
};

const validateCarBrandField = (text: string): string => {
    let res: string = '';
    const carBrandRegexp = /[\p{Script=Latin}\p{White_Space}]/gu;
    const carBrandLengthRules: TextLengthRules = {
        minChars: 4,
        maxChars: 20,
    };
    const carBrandTextRules: TextRules = {
        lengthRules: carBrandLengthRules,
        regexp: carBrandRegexp,
    };
    if (!textFollowsRules(text, carBrandTextRules)) {
        res = 'Только латинские буквы от 4х до 20 символов';
    }
    return res;
};

const validateEmailField = (text: string): string => {
    let res: string = '';
    const emailRules: TextRules = {
        regexp: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/g,
    };
    if (!textFollowsRules(text, emailRules)) {
        res = 'Пожалуйста, заполните это поле';
    }
    return res;
};

const filterTextFromSpecificCharacters = (text: string): string => {
    const normalTextRegexp =
        /([\p{Alphabetic}\p{Mark}\p{Decimal_Number}\p{Punctuation}\p{Join_Control}]+)/gu;
    return text.match(normalTextRegexp)?.join(' ') as string;
};

const validateTextField = (text: string): string => {
    let res: string = '';
    if (!filterTextFromSpecificCharacters(text.trim()).length) {
        res = 'Пожалуйста, заполните это поле';
    }
    return res;
};

export const validateForm = (form: Question): FormValidationProps => {
    const res: FormValidationProps = {
        errors: getEmptyQuestionObject(),
        isValid: true,
    };
    const { user, text, carBrand, email, carModel } = form;

    res.errors = {
        user: validateUserField(user),
        carModel: validateCarModelField(carModel),
        carBrand: validateCarBrandField(carBrand),
        email: validateEmailField(email),
        text: validateTextField(text),
    };

    if (!isNotEmpty(res.errors)) res.isValid = false;
    return res;
};

const getFormattedDate = (): string => {
    const today = new Date();
    const curDay = today.getDay() < 10 ? '0' + today.getDay() : today.getDay();
    const curMonth =
        today.getMonth() + 1 < 10 ? '0' + (today.getMonth() + 1) : today.getMonth() + 1;
    const ddmmyyyy = curDay + '.' + curMonth + '.' + today.getFullYear();
    const time = today.getHours() + ':' + today.getMinutes();
    return ddmmyyyy + 'T' + time;
};

export const preparePayload = (data: Question): Question => {
    const dateTime = getFormattedDate();
    return {
        ...data,
        text: filterTextFromSpecificCharacters(data.text),
        date: dateTime,
    };
};
