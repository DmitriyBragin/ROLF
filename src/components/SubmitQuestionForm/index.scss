.submit-question-form {
    width: 50%;
    @media screen and (max-width: 500px) {
        width: 100%;
    }
}

.MuiFormHelperText-root {
    text-align: right !important;
    margin: 0 !important;
}
