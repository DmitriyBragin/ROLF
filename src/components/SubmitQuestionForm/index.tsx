import './index.scss';
import { Question } from '../../utils/types';
import { Box, Button, Grid, TextField } from '@mui/material';
import React, { useEffect, useState } from 'react';
import { getEmptyQuestionObject, isFilled, preparePayload, validateForm } from './helpers';

export const SubmitQuestionForm = ({ cb }: { cb: (q: Question) => void }) => {
    const [formErrors, setFormErrors] = useState<Question>(getEmptyQuestionObject());
    const [data, setData] = useState<Question>(getEmptyQuestionObject());
    const [canSubmit, setCanSubmit] = useState<boolean>(false);

    const handleChange = (evt: { target: any }) => {
        setData({
            ...data,
            [evt.target.name]: evt.target.value,
        });
    };

    useEffect(() => {
        setCanSubmit(isFilled(data));
    }, [JSON.stringify(data)]);

    const cleanupData = (): void => {
        setData(getEmptyQuestionObject());
        setFormErrors(getEmptyQuestionObject());
    };

    const handleSubmit = (e: any) => {
        e.preventDefault();
        const { isValid, errors } = validateForm(data);
        setFormErrors(errors);
        if (isValid) {
            const payload: Question = preparePayload(data);
            cleanupData();
            cb(payload);
        }
    };

    return (
        <div className="submit-question-form">
            <Box component="form" onSubmit={handleSubmit} sx={{ mt: 3 }}>
                <Grid container spacing={2}>
                    <Grid item xs={12} sm={6}>
                        <TextField
                            name="user"
                            required
                            fullWidth
                            id="user"
                            label="Имя"
                            value={data.user}
                            onChange={handleChange}
                            error={!!formErrors.user.length}
                            helperText={formErrors.user}
                        />
                    </Grid>
                    <Grid item xs={12} sm={6}>
                        <TextField
                            required
                            fullWidth
                            id="email"
                            label="Email"
                            name="email"
                            value={data.email}
                            error={!!formErrors.email.length}
                            helperText={formErrors.email}
                            onChange={handleChange}
                        />
                    </Grid>
                    <Grid item xs={12} sm={6}>
                        <TextField
                            required
                            fullWidth
                            id="carBrand"
                            label="Марка авто"
                            name="carBrand"
                            value={data.carBrand}
                            error={!!formErrors.carBrand.length}
                            helperText={formErrors.carBrand}
                            onChange={handleChange}
                        />
                    </Grid>
                    <Grid item xs={12} sm={6}>
                        <TextField
                            required
                            fullWidth
                            id="carModel"
                            label="Модель авто"
                            name="carModel"
                            value={data.carModel}
                            error={!!formErrors.carModel.length}
                            helperText={formErrors.carModel}
                            onChange={handleChange}
                        />
                    </Grid>
                    <Grid item xs={12}>
                        <TextField
                            multiline
                            required
                            fullWidth
                            name="text"
                            label="Введите текст вопроса"
                            type="text"
                            id="text"
                            value={data.text}
                            inputProps={{ maxLength: 500 }}
                            error={!!formErrors.text.length}
                            helperText={`${formErrors.text} ${data.text.length}/500`}
                            onChange={handleChange}
                        />
                    </Grid>
                </Grid>
                <Button
                    type="submit"
                    variant="contained"
                    sx={{ mt: 3, mb: 2 }}
                    disabled={!canSubmit}>
                    Отправить
                </Button>
            </Box>
        </div>
    );
};
