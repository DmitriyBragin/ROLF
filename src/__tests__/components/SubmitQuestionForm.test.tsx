import { render } from '@testing-library/react';
import React from 'react';

import { SubmitQuestionForm } from '../../components/SubmitQuestionForm';

describe('SubmitQuestionForm', () => {
    test('renders', () => {
        const { container } = render(<SubmitQuestionForm cb={() => {}} />);
        expect(container).toMatchSnapshot();
    });
});
