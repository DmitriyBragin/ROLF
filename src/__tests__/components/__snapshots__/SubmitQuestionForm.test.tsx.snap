// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitQuestionForm renders 1`] = `
<div>
  <div
    class="submit-question-form"
  >
    <form
      class="MuiBox-root css-178yklu"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="user"
              id="user-label"
            >
              Имя
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="user"
                name="user"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1g8cwq9"
                >
                  <span>
                    Имя
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="email"
              id="email-label"
            >
              Email
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="email"
                name="email"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1g8cwq9"
                >
                  <span>
                    Email
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="carBrand"
              id="carBrand-label"
            >
              Марка авто
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="carBrand"
                name="carBrand"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1g8cwq9"
                >
                  <span>
                    Марка авто
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="carModel"
              id="carModel-label"
            >
              Модель авто
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="carModel"
                name="carModel"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1g8cwq9"
                >
                  <span>
                    Модель авто
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="text"
              id="text-label"
            >
              Введите текст вопроса
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-describedby="text-helper-text"
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                id="text"
                maxlength="500"
                name="text"
                required=""
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1g8cwq9"
                >
                  <span>
                    Введите текст вопроса
                     
                    *
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
              id="text-helper-text"
            >
               0/500
            </p>
          </div>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-1n7zltw-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        Отправить
      </button>
    </form>
  </div>
</div>
`;
